File: Codes in C++/Dynamic Programming/Best Time to Buy and Sell Stock with Cooldown.cpp
// Space Optimised
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<int> front2(2, 0);
        vector<int> front1(2, 0);
        vector<int> curr(2, 0);
        for (int ind = n - 1; ind >= 0; ind--) {
            curr[1] = max(-prices[ind] + front1[0], 0 + front1[1]);
            curr[0] = max(prices[ind] + front2[1], 0 + front1[0]);
            front2=front1;
            front1=curr;
        }
        return front1[1];
    }
}; 


// class Solution {
// public:
//     int maxProfit(vector<int>& prices) {
//         int n = prices.size();
//         vector<vector<int>> dp(n + 2, vector<int>(2, 0));
//         for (int ind = n - 1; ind >= 0; ind--) {
//             // for(int buy=0;buy<=1;buy++)
//             // {
//             //     if(buy)
//             //     {
//             //         dp[ind][buy]=max(-prices[ind]+dp[ind+1][0],0+dp[ind+1][1]);
//             //     }
//             //     else{
//             //         dp[ind][buy]=max(prices[ind]+dp[ind+2][1],0+dp[ind+1][0]);
//             //     }
//             // }

//             dp[ind][1] = max(-prices[ind] + dp[ind + 1][0], 0 + dp[ind + 1][1]);
//             dp[ind][0] = max(prices[ind] + dp[ind + 2][1], 0 + dp[ind + 1][0]);
//         }
//         return dp[0][1];
//     }
// };

File: Codes in Python3/Dynamic Programming/Best Time to Buy and Sell Stock with Cooldown.py
#Space Optimised Bottom Up Approach
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n=len(prices)
        next1=[0]*2 
        next2=[0]*2
        for i in range(n-1,-1,-1):
            curr=[0]*2 
            for buy in range(1,-1,-1):
                if buy:
                    curr[buy]=max(next1[1],-prices[i]+next1[0]) 
                else:
                    curr[buy]=max(next1[0],next2[1]+prices[i])             
            next2=next1
            next1=curr
        return next1[1]

# TC=O(n)
# SC=O(1)
        

#Bottom Up Approach
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n=len(prices)
        dp=[[0]*2 for _ in range(n+2)]
        for i in range(n-1,-1,-1):
            for buy in range(1,-1,-1):
                if buy:
                    dp[i][buy]=max(dp[i+1][1],-prices[i]+dp[i+1][0]) 
                else:
                    dp[i][buy]=max(dp[i+1][0],dp[i+2][1]+prices[i])             
       
        return dp[0][1]

# TC=O(n)
# SC=O(n)
        
#Top Down Approach
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n=len(prices)
        dp=[[-1]*2 for _ in range(n)]
        def f(i,buy):
            if i>=n:
                return 0
            if dp[i][buy]!=-1:
                return dp[i][buy]
            if buy:
                dp[i][buy]=max(f(i+1,1),-prices[i]+f(i+1,0))
            else:
                dp[i][buy]=max(f(i+1,0),f(i+2,1)+prices[i])
            
            return dp[i][buy]
        return f(0,1)

# TC=O(n)
# SC=O(n)
        

#Top Down Approach
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n=len(prices)
        dp=[[-1]*2 for _ in range(n)]
        def f(i,buy):
            if i>=n:
                return 0
            if dp[i][buy]!=-1:
                return dp[i][buy]
            if buy:
                dp[i][buy]=max(f(i+1,1),-prices[i]+f(i+1,0))
            else:
                dp[i][buy]=max(f(i+1,0),f(i+2,1)+prices[i])
            
            return dp[i][buy]
        return f(0,1)

# TC=O(n)
# SC=O(n)
        